src/storage/src/sink/kafka.rs

```diff
@@ -881,33 +882,43 @@ impl KafkaSinkState {
     /// Updates the latest progress update timestamp based on the given
     /// input frontier and pending rows.
     ///
-    /// This will emit an `END` record to the progress topic if the frontier
+    /// This will emit a progress record to the progress topic if the frontier
     /// advanced and advance the maintained write frontier, which will in turn
     /// unblock compaction of timestamp bindings in sources.
     ///
-    /// *NOTE*: `END` records will only be emitted when
+    /// *NOTE*: Progress records will only be emitted when
     /// `KafkaSinkConnection.consistency` points to a progress topic. The
     /// write frontier will be advanced regardless.
     async fn maybe_emit_progress(
         &mut self,
-        input_frontier: AntichainRef<'_, Timestamp>,
+        mut input_frontier: Antichain<Timestamp>,
+        as_of: &SinkAsOf<Timestamp>,
     ) -> anyhow::Result<bool> {
+        input_frontier.extend(self.pending_rows.keys().min().cloned());
+        let min_frontier = input_frontier;
+
+        // If we emit a progress record before the as_of, we open ourselves to the possibility that
+        // we restart the sink with a gate timestamp behind the ASOF.  If that happens, we're unable
+        // to tell the difference between some records we've already written out and those that we
+        // still need to write out. (This is because asking for records with a given ASOF will fast
+        // forward all records at or before to the requested ASOF.)
+        if !PartialOrder::less_equal(&as_of.frontier, &min_frontier) {
```

Comment:
In `2702130f950533ce9e5cdbf713d2cae6402411aa`, I moved this comparison so it includes `self.pending_rows.keys().min()`.  That isn't strictly necessary since that value can never be before the ASOF -- but I think it's much more clear this way so you don't have to think about that invariant when reading this code

Reply:
Cool. I believe this is safe because:
- The antichain extend takes the min timestamp, so the value of `min_frontier` has not changed.
- The `if` condition here will be true whenever the old condition was true.

Reply:
Chae and I just discussed, we think it should be `less_than`, and not `less_equal` because we subtract `1` below when calculating the timestamp that we emit.

Reply:
Interesting! It sounds like this was another lurking bug then?